woocommerce/woocommerce-gutenberg-products-block
# Coupon and Cart Item Validation Fixes

This PR has several fixes related to validation on the cart and checkout. It resolves the following bug reports:

_No validation when you created an email restricted coupon #2235_

Fixed by running late validation before payment. Payment is prevented, and restricted coupons are removed. This updates cart totals so the payment can be attempted again.

Fixes #2235

_Ensure our Add to Cart has parity with Core Add to Cart in term of filters #2218_

We were missing 2 general purpose filters from woo core that plugins use to run custom validation:

- `woocommerce_add_to_cart_validation`
- `woocommerce_check_cart_items`

I've added them to the API and included our Notice->Exception logic to avoid notices being added in the background.

Fixes #2218

_Coupons with restrictions are showing raw errors (HTML) when restrictions are not met. #2212_

All we get from core is a message; this is resolved by stripping HTML tags before providing the API response.

Fixes #2212

_Previously applied coupons that no longer meet criteria are still visible in cart (even if they do nothing) #2213_

We have logic in place to gather errors for cart items, so I added similar logic for coupons. The coupon is removed, and a notice is surfaced. To keep notices visible when doing lots of cart updates, I made it dismissable and prevented all notice being cleared after cart updates.

#2213

### How to test the changes in this Pull Request:

**Test case 1:**
1. Create a coupon with a min spend
2. Apply min spend coupon without meeting requirements. Check error message is correctly formatted.
3. Meet requirements by adding products to cart. Add coupon again.
4. Reduce cart qty until requires are not met again. Check notice is shown and the coupon is removed.

**Test case 2:**
1. Create a coupon with a usage restriction of `*@automattic.com`
2. Apply it.
3. Go to place an order without setting the billing email to an automattic address.
4. See error notice and cart removal. Cart totals should be correct.

**Test case 3:**
1. Create a product with a dependency using https://en-gb.wordpress.org/plugins/woocommerce-product-dependencies/
2. Try to add it to cart. Add to cart will fail. There is no notice because of https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2242
